mtp_cashbook/apps/disbursements/views.py

```diff
@@ -576,19 +593,30 @@ def get_context_data(self, **kwargs):
         return context
 
     def form_valid(self, form):
-        update = self.get_update_payload(form)
-        changed_fields = set()
-        for field in update:
-            if update[field] != self.disbursement[field]:
-                changed_fields.add(field)
-        if 'prisoner_number' in changed_fields:
-            changed_fields.add('prison')
-        if changed_fields:
-            if 'remittance' in changed_fields:
-                changed_fields.remove('remittance')
+        disbursement_data = self.get_update_payload(form)
+        if disbursement_data.get('recipient_type') == 'person':
+            disbursement_data['recipient_is_company'] = False
+        elif disbursement_data.get('recipient_type') == 'company':
+            disbursement_data['recipient_is_company'] = True
+            disbursement_data['recipient_last_name'] = disbursement_data['recipient_company_name']
+            disbursement_data['recipient_first_name'] = ''
+        if 'recipient_type' in disbursement_data:
+            del disbursement_data['recipient_type']
+        if 'recipient_company_name' in disbursement_data:
+            del disbursement_data['recipient_company_name']
+        if 'remittance' in disbursement_data:
+            del disbursement_data['remittance']
```

Comment:
I would think all this stuff should go in the specific `get_upload_payload` of the form or view, rather than in the base class.